### PR TITLE
Restore -vmthreads flag as an error

### PR DESCRIPTION
**Pull request by dra27**
I'm sorry, I didn't look at #2289 in time to comment there. Apropos https://github.com/ocaml/ocaml/pull/2276#issuecomment-470865570, however unused it may be, I don't think that a CLI flag should simply disappear.

This GPR restores `-vmthread` with an adapted version of the deprecation message as an error message and also keeps the `use_vmthreads` part of ppx contexts.

I realise how pointless this looks, but experience of firefighting breakage around removing/renaming flags and things in the opam, opam-depext and various other parts of that "which no one uses" has left me highly suspicious of removing flags or variables ever!

### CODE REVIEW

**Comment by avsm** [comment]
I totally agree; this will make finding older uses of vmthreads much much easier!

**Comment by ghost** [comment]
That seems fine to me if it helps package maintainers. When do you plan to remove these completely?

**Comment by dra27** [comment]
@diml - never, well, or perhaps when the driver is completely overhauled. I can't conceive how this particular one would ever come back into being, but in the case of `-plugin` in 2276, it also serves as a permanent code marker that "this flag meant something once, consider that if you're about to re-use it".

File: driver/main.ml [hunk 1]
```diff
@@ -102,6 +102,7 @@ module Options = Main_args.Make_bytecomp_options (struct
   let _strict_formats = set strict_formats
   let _no_strict_formats = unset strict_formats
   let _thread = set use_threads
+  let _vmthread = set use_vmthreads
```
Comment (xavierleroy) [inline driver/main.ml]: Why not failing immediately when reading the `-vmthread` argument?

File: utils/clflags.ml [hunk 2]
```diff
@@ -68,6 +68,7 @@ let absname = ref false                 (* -absname *)
 let annotations = ref false             (* -annot *)
 let binary_annotations = ref false      (* -annot *)
 and use_threads = ref false             (* -thread *)
+and use_vmthreads = ref false           (* -vmthread *)
```
Comment (xavierleroy) [inline utils/clflags.ml]: I would really like to get rid of this `use_vmthreads` reference.

**Review by xavierleroy — COMMENTED** [review]
The intent is good but I don't understand the way it is implemented, see below.  Can you explain?

**Comment by shindere** [comment]
Xavier Leroy (2019/03/12 16:12 +0000):
> Why not failing immediately when reading the `-vmthread` argument?

I agree it should happen like this.

> > @@ -68,6 +68,7 @@ let absname = ref false                 (* -absname *)
>  let annotations = ref false             (* -annot *)
>  let binary_annotations = ref false      (* -annot *)
>  and use_threads = ref false             (* -thread *)
> +and use_vmthreads = ref false           (* -vmthread *)
> 
> I would really like to get rid of this `use_vmthreads` reference.

Indeed!

**Comment by dra27** [comment]
@xavierleroy - the "implementation" is to revert the change from #2289, so the reference is simply that which was in `Clflags` before. That's no particular defence, except that there's basically no new code relative to the pre-#2289 version except to change deprecated to fatal.

I don't have a particularly strong opinion, but I believe that the arguments parser is supposed to parse the arguments and then report on any semantic issues - as this GPR stands, and in present releases, `ocamlc -vmthread -foo` complains about not understanding what `-foo` is, whereas what you propose displays the removal of vmthreads error message. More confusingly, it would mean that `ocamlc -foo -vmthread` and `ocamlc -vmthread -foo` display different error messages (however correct each may be).

(Not that I'm in favour of the myriad references in `Clflags`, of course!)

**Comment by xavierleroy** [comment]
Dear David,

With all due respect I think you're overengineering it :-)  It is no problem that `ocamlc -foo -vmthread` and `ocamlc -vmthread -foo` display different error messages, as long as they display error messages, because they are now both erroneous.  No need for a distinction between "syntactic" and "semantic" analysis of command-line flags, a distinction that is not obvious to me to begin with.

On the other hand, I feel strongly about keeping  `Clflags` "clean", i.e. all references should correspond to actively-supported options.  So, `Clflags.use_vmthreads` has to go.

**Comment by dra27** [comment]
It is the first time I have overengineered something by *not* writing code 🙂

**Review by xavierleroy — APPROVED** [review]
Looks fine to me.  Thank you for the changes.

**Pull request merged**